Try auto-detecting CUDA compute capability

masterleinad opened the pull request:
Fixes #2455. Doing the detection in this way worked pretty well for another project.
To avoid redeclaring the `CUDA` architecture `CMake` options, this pull request also introduces a `kokkos_set_option` function that relies on storing previously registered options in lists simulating maps in `CMake`.

dalg24 commented on cmake/kokkos_arch.cmake:
```diff
@@ -389,7 +396,34 @@ IF (KOKKOS_ENABLE_OPENMPTARGET)
+    ELSE()
+      MESSAGE(STATUS "Detected CUDA Compute Capability ${CUDA_COMPUTE_CAPABILITY}")
+      LIST(GET KOKKOS_CUDA_ARCH_LIST ${FLAG_INDEX} ARCHITECTURE)
+      KOKKOS_SET_OPTION(${ARCHITECTURE} ON)
```
Why did you introduce that function?  Is there an issue with doing
```CMake
SET(KOKKOS_ARCH_${ARCHITECTURE} ON)
```

masterleinad replied:
The main objective of the function is to set `Kokkos_ARCH_$ARCHITECTURE` so that it is user-visible what we detected. Since it is a `CACHE` variable, we need to `FORCE` it and overwrite all the previous settings for the option (including the docstring).

jrmadsen replied:
So I have a couple of functions that sort fixes this issue. Whenever you are potentially changing a variable locally, you do `add_feature(<OPTION> <DOCSTRING>)` in CMake and then you just print out those values at the end of CMake. 

```console
-- The following features are defined/enabled (+):
     TIMEMORY_USE_CALIPER: Enable Caliper
     TIMEMORY_USE_CUDA: Enable CUDA option for GPU measurements
     TIMEMORY_USE_DYNINST: Enable dynamic instrumentation

-- The following features are NOT defined/enabled (-):
     TIMEMORY_USE_ALLINEA_MAP: Enable control for AllineaMAP sampler
     TIMEMORY_USE_ARCH: Enable architecture flags
     TIMEMORY_USE_CLANG_TIDY: Enable running clang-tidy
```

dalg24 replied:
I assume you mean `add_feature_info()`

masterleinad replied:
I think it's https://github.com/NERSC/timemory/blob/5c0f89880db25e31a0778d8252b05c4ad5e50803/cmake/Modules/MacroUtilities.cmake#L690-L707.

For me, this feels a little out-of-scope of this pull request. In the end, I care about what is in the cache. Of course, we can discuss printing a summary of the enabled and disabled features at the end of the `CMake` run separately.

dalg24 reviewed (dismissed):
Please provide a rational for `kokkos_set_option()`

masterleinad commented:
OK. So it doesn't work with `clang` because we likely need extra flags. Since this is only an attempt to auto-detect and we fail in a reasonable way, I would still consider this good enough for now (once everyone is happy with the `CMake` here).

masterleinad commented on cmake/kokkos_arch.cmake:
```diff
@@ -389,7 +396,35 @@ IF (KOKKOS_ENABLE_OPENMPTARGET)
+      CHECK_CUDA_ARCH(${ARCHITECTURE} sm_${CUDA_COMPUTE_CAPABILITY})
+    ENDIF()
+  ELSE()
+    MESSAGE(SEND_ERROR "CUDA enabled but no NVIDIA GPU architecture currently enabled and auto-detection failed. Please give one -DKokkos_ARCH_{..}=ON' to enable an NVIDIA GPU architecture.")
```
> Is there a reason we do this with two execute_process calls instead of try_run?

This is what I had already. I can have a look to use `try_run` instead if you prefer.

> The most likely reason auto-detection fails is compiling on a login node. Is there a good way to 1) tell them to try configuring from a compute node? or 2) make the utility easy to build and run?
 
We can mention that in the error message. In the end, the source file should be stand-alone anyway.

masterleinad commented:
Retest this please.

masterleinad commented:
@jjwilke Can you please comment on me defining `kokkos_set_option` for being able to set the value of a `CACHE` variable at a later point than when we define it?
Do you think we should do that (to make the user aware of what we changed also in `CMakeCache.txt` and to avoid to rerun the check when reconfiguring)? Do you see a better approach do achieve the same?

jjwilke commented on cmake/kokkos_functions.cmake:
```diff
@@ -73,7 +80,20 @@ FUNCTION(kokkos_option CAMEL_SUFFIX DEFAULT TYPE DOCSTRING)
 
+  list(GET KOKKOS_OPTION_VALUES ${OPTION_INDEX} DOCSTRING)
+  list(GET KOKKOS_OPTION_TYPES ${OPTION_INDEX} TYPE)
+  SET(${CAMEL_NAME} ${VALUE} CACHE ${TYPE} ${DOCSTRING} FORCE)
```
I think this probably makes sense for this case. I'm okay doing a cache override. I would at least print a STATUS message about what's going on.

If we don't set the cache variable, then I think what makes the most sense is error-ing out.  Print the CUDA capability detected, but force the user to specify it.

masterleinad commented:
I think that I addressed all review comments. The executable can be compiled in `SM_ONLY` mode which we use inside the auto-detection. Otherwise, we print the `CMake` option to set. Sample output:
```
Set -DKokkos_ARCH_VOLTA70=ON .`
```
`set_option` now also prints if it does something.

jjwilke commented:
This looks good. I'm ready to approve. Thanks for doing this.  I hate to bike shed - but should we squash? We have 15 commits, many of which seem like intermediate commits.

masterleinad commented:
Oh, sure I am happy to squash once everyone is happy.

masterleinad committed:
Introduce kokkos_set_option

masterleinad committed:
Allow auto-detecting the CUDA architecture

masterleinad committed:
Auto-detect CUDA architecture in CI

masterleinad commented:
@jjwilke I rebased and squashed.

masterleinad commented:
Retest this please.

dalg24 commented:
@masterleinad Did you edit my commit?  I see https://github.com/kokkos/kokkos/blob/4bf7168587b450d3c9e89a2f85dac8a4b1a1039c/.jenkins#L66 is back

masterleinad commented:
> @masterleinad Did you edit my commit?

After saying that it doesn't work for `clang` yet in https://github.com/kokkos/kokkos/pull/2980#issuecomment-620728131, I explicitly passed the architecture flag back for the `clang+cuda` build. I didn't modify anything while squashing.

masterleinad commented:
I am sorry if I haven't made it clear enough that `clang` requires extra flags that we don't try to pass here.

jrmadsen commented on cmake/kokkos_functions.cmake:
```diff
@@ -47,6 +47,13 @@ FUNCTION(kokkos_option CAMEL_SUFFIX DEFAULT TYPE DOCSTRING)
   STRING(TOUPPER ${CAMEL_NAME} UC_NAME)
 
+  LIST(APPEND KOKKOS_OPTION_KEYS ${CAMEL_SUFFIX})
+  SET(KOKKOS_OPTION_KEYS ${KOKKOS_OPTION_KEYS} PARENT_SCOPE)
```
I think a global or directory property would be better suited here.
